Fix: skewed placeholder of product image - issue#7553

kmanijak opened the pull request:
### Summary
Fix for the [issue #7553](https://github.com/woocommerce/woocommerce-blocks/issues/7553).

Placeholder image was skewed (during loading and once loaded) when there was not enough space for it to expand, for example:
- when used in a Columns block,
- on a narrow screen.

What wasn't mention in the issue is that the issue could be reproduced on too wide screens as well (when single product had more than 500px width).

#### Cause
The cause of the glitch were inline styles on the placeholder image in `Product Image` block:
```
    <img src={ PLACEHOLDER_IMG_SRC } alt="" width={ 500 } height={ 500 } />
```
`width` was overridden by more specific styles, while `height` was not, which made the placeholder skewed with the height fixed to 500px.

#### Changes
- Removal of the inline styles let the placeholder to keep the image ratio depends on the width. That works in both cases: when there's not enough or too much space for a product.

- Additionally, the corresponding styles where removed from a PHP code. [First commit](https://github.com/woocommerce/woocommerce-blocks/commit/5c10b403da153546c1368ad57a604e6393ff7220) shows that the styles were anyway not even applied, which proves it's safe to remove them. 

- Fix a couple of TS errors in tests: [commit](https://github.com/woocommerce/woocommerce-blocks/pull/7651/commits/4763d7f182198077b06bfaa2844a6a4227ccdd06)

- Add explicit comments about alt text being empty in decorative images: [commit](https://github.com/woocommerce/woocommerce-blocks/pull/7651/commits/a8d13c83b73056d2bcb360134c9c7630d41a09ac)

#### Fixes:
- [issue #7553](https://github.com/woocommerce/woocommerce-blocks/issues/7553)

### Screenshots

| Case | Before | After |
| ------ | ----- | ----- |
|     Single product (wider than 500px)   |    <img width="1224" alt="image" src="https://user-images.githubusercontent.com/20098064/201080282-4ea921d5-732d-41ae-97c6-6f33316ed51f.png">  |    <img width="1244" alt="image" src="https://user-images.githubusercontent.com/20098064/201080761-3d8172d2-bacb-4b51-b1f7-1ca6db7f82aa.png">   |
|     Single product (narrower than 500px)   |    <img width="591" alt="image" src="https://user-images.githubusercontent.com/20098064/201081115-ec343644-df45-48b3-8458-9c2b670d4ac8.png">  |    <img width="587" alt="image" src="https://user-images.githubusercontent.com/20098064/201080889-3a6bf5c7-e7d4-42a9-b0c2-5d0c84189b1e.png">   |
|     Product Query / All Products   |     <img width="571" alt="image" src="https://user-images.githubusercontent.com/20098064/201080546-598f2b20-da27-439e-a278-64f209d50e66.png">   |    <img width="569" alt="image" src="https://user-images.githubusercontent.com/20098064/201080979-8c854d30-98bd-4863-92b6-cfac48a3f8db.png">   |

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

Prerequisites:
0. Make sure you have a product without a highlighted image (you can create a new product or remove the image from an existing one).

Case 1: image is wider than 500px
1. Go to Appearance > Site Editor > Templates > Product Catalog.
2. Remove the legacy placeholder block.
3. Add blocks that use Product Image placeholder:
- Single Product
- All Products
- Product Query
4. ✅ Check on the wide-screen (image should have more than 500px width) if the placeholder image is a square (not skewed)

Case 2: image is narrower than 500px
1. Add a Columns block.
2. Inside one of the columns, add the Single Product block, to another one add Product Query block. Columns block limits the available space for a product to display (you should achieve less than 500px width for product image)
3. ✅ Check if the placeholder image is a square (not skewed)

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Fix skewed placeholder of a Product Image block

Commit:
Fix typo in HTML markup of ProductImage block placeholder

Typo caused rendering of incorrect HTML attributes: width and height of Product Image placeholder that were anyway ignored by the browser

kmanijak committed:
Remove inline styles (width and height) from ImagePlaceholder in ProductElements > Image block

Inline height took precedence over the inherited styles which made the placeholder image skewed (in loading and loaded state). Removal of those styles allows the ImagePlaceholder to adapt the height to the available space keeping the ratio or inherit the styles from the parent

kmanijak committed:
Remove inline styles (width and height) from placeholder image in ProductImage.php block

Inline styles applied to the placeholder image of ProductImage block were overriden by other styles with higher specificity, which made them redundant. Additionally, corresponding styles were removed from the placeholder image from Editor code as they caused UI glitches. Additional proof that it's safe to remove them is in the first commit in this branch, the purpose of which was to fix those styles as there was a typo which corrupted them and eventually inline width and height were ignored by the browser and not applied to the element

kmanijak committed:
Add test to check placeholder image renders without width and height inline attributes

This is to prevent adding inline width and height attributes so the sizing of the placeholder image is controlled by the inherited styles or element styles, in the same way as a regular product image

github-actions[bot] committed:
bot: update checkstyle.xml

kmanijak committed:
Fix TypeScript errors in the block test file of Product Image

github-actions[bot] committed:
bot: update checkstyle.xml

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/test/block.test.js:
```diff
@@ -62,7 +62,10 @@ describe( 'Product Image Block', () => {
-				<ProductDataContextProvider product={ productWithImages }>
+				<ProductDataContextProvider
+					product={ productWithImages }
+					isLoading={ false }
```
Can you clarify why `isLoading` is hardcoded to `false` in the context of this test, and why was it needed to be added as a fix for this issue? 🤔

kmanijak replied:
> Can you clarify why isLoading is hardcoded to false in the context of this test

Sure! I added `isLoading` prop to each test, as I realised it's a required prop for `ProductDataContextProvider` and TypeScript checker complained about that.
And considering the tests are checking the behaviour of component with images loaded already, I checked the code and `isLoading` passed as `false` seems to set such condition. Also, there's no default value for `isLoading`, so before adding that to tests, it was implicitly falsy.
But I may be missing something, so I'm all ears if you see that may be troublesome! 👂 

> and why was it needed to be added as a fix for this issue? 🤔

TS fixes/improvements were added as an extra to this PR and are not required as a fix for issue. I gathered them [in a separate commit](https://github.com/woocommerce/woocommerce-blocks/pull/7651/commits/4763d7f182198077b06bfaa2844a6a4227ccdd06), but maybe I should've added it as a separate PR to avoid confusion. Let me know the way forward 🙌 
EDIT: Or add it to the PR description in "Changes" section. Will keep that in mind for future.

sunyatasattva reviewed (commented):
I've got just a small question, otherwise it looks amazing! Great PR description!

thealexandrelara commented on assets/js/atomic/blocks/product-elements/image/block.js:
```diff
@@ -129,9 +129,7 @@ export const Block = ( props ) => {
-	return (
-		<img src={ PLACEHOLDER_IMG_SRC } alt="" width={ 500 } height={ 500 } />
-	);
+	return <img src={ PLACEHOLDER_IMG_SRC } alt="" />;
```
I saw that `alt` prop was already empty before, but I'm wondering if it would be better to add some text to explicitly say it is a placeholder image instead. What do you think?

sunyatasattva replied:
Super-duper agreed!

kmanijak replied:
Thank you for this suggestion! I was thinking about that, but then in a test I spotted the placeholder is searched by the empty alt:

```const placeholderImage = component.getByAltText( '' );```

and assumed it may be empty for a reason. But two `aye!`s are enough to convince me! 💪 

I'll double-check there's no other logic based on that and provide it.

kmanijak replied:
Alt text added [in the commit](https://github.com/woocommerce/woocommerce-blocks/pull/7651/commits/826f1d1c96ccb05f13d99f771380643e63e6b5fc).

I considered what should be the alt text:
- what's really on the image, like "Simple landscape drawing"
- what's the purpose, like "Generic placeholder", since in this case it doesn't really matter what's there
- something else

I went for a second option since the image itself is set externally through the settings and may change over time. Also, the content is kind of redundant in this context, the purpose matters. There are even opinions that placeholder image doesn't need alt text at all (https://www.seroundtable.com/google-alt-text-small-images-23672.html - thanks @imanish003 for sharing that! 🙌 ), but I'm not convinced.

Happy to hear your opinion!

thealexandrelara replied:
Great work @kmanijak. I think it is great that @imanish003 mentioned the fact that placeholder images may not need an alt text and yes I just remembered that are some situations in which the alt text must be empty, but I cannot remember from the top of my mind all those situations so I went ahead looking for resources related to that.

I found this alt Decision Tree from the W3C Web Accessibility Initiative (WAI) that I think we can use to make this decision: https://www.w3.org/WAI/tutorials/images/decision-tree/. It seems that if we consider the placeholder image as being just **decorative or not intended for the user** the best solution would be to leave it empty. Here is also a resource talking about Decorative Images: https://www.w3.org/WAI/tutorials/images/decorative/, it states that "Text values for these types of images would add audible clutter to screen reader output or could distract users if the topic is different from that in adjacent text", which is something I can agree since we can have dozens of placeholder images that is going to be read by the screen readers.

sunyatasattva replied:
This was a very interesting discussion. After reading it, I agree that alt text would add audible clutter to screen reader users. People do not need to hear over and over about the placeholder image which is there just to keep the layout, actually. Does not add any information.

I agree with removing it. Sorry @kmanijak for the extra work. But it's a good opportunity also because I saw there were some mistakes in the implementation which we take for granted and it's a good time to inform you about them. We never use naked strings in either JS or PHP code, we always use WordPress localization functions.

Here are two resources:

* https://codex.wordpress.org/I18n_for_WordPress_Developers
* https://www.npmjs.com/package/@wordpress/i18n

Most generally, you will see that you'll often find yourself using the `__` function.


With that said, I think we should make this decision of leaving the alt text empty explicit by adding a comment in the relevant places, perhaps linking to this discussion thread. Someone else could see the empty alt text and it would be the same story all over again!

kmanijak replied:
Thank you @thealexandrelara and @sunyatasattva for your valuable input! I'm happy with the conclusion.

The changes are available in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/7651/commits/a8d13c83b73056d2bcb360134c9c7630d41a09ac):
- remove alt text (especially as an explicit string, thanks Lucio for pointing this out, I'll keep that in mind!)
- add explicit comments to all the relevant places I found.

kmanijak committed:
Add generic alt text to placeholder image

Alt text added in this commit is a generic text, not description of the actual image. That's because the image itself is set externally through the settings and may change over time

kmanijak commented:
@sunyatasattva, I responded to your comments. Happy to hear your opinion regarding `alt` text of the placeholder image!

sunyatasattva reviewed (changes requested):
Left a comment in the `alt` text thread. Sorry for the back and forth.

kmanijak committed:
Revert adding placeholder image alt text and add comments to make the empty alt explicit

After a Github discussion: https://github.com/woocommerce/woocommerce-blocks/pull/7651\#discussion_r1019560494 it was decided the placeholder should NOT have alt text as it serves the purpose of decorative image. According to the guidelines decorative images should not have alt text: https://www.w3.org/WAI/tutorials/images/decorative/. Comments were added also to other places where it happens

github-actions[bot] committed:
bot: update checkstyle.xml

kmanijak commented:
No worries @sunyatasattva, I'm happy we came up with a proper conclusion 🙌 

- I updated the `alt` text and add explicit comments as suggested
- left the comment in `alt` discussion
- left the comment in [`isLoading` discussion](https://github.com/woocommerce/woocommerce-blocks/pull/7651#discussion_r1019514015)

I'd appreciate another round of review.

kmanijak committed:
Merge branch 'trunk' into fix/placeholder-product-image-is-skewed

github-actions[bot] committed:
bot: update checkstyle.xml

thealexandrelara reviewed (approved):
Great work! LGTM!

sunyatasattva reviewed (approved):
Excellent! :shipit: !

kmanijak committed:
Merge branch 'trunk' into fix/placeholder-product-image-is-skewed
